File: backend/app/schemas/transfers.py
"""Esquemas para el módulo de transferencias entre sucursales."""
from __future__ import annotations

from datetime import datetime
from decimal import Decimal

from pydantic import BaseModel, ConfigDict, Field, AliasChoices, field_validator

from backend.app.models import TransferStatus
from .audit import AuditTrailInfo


class TransferOrderItemBase(BaseModel):
    device_id: int = Field(..., ge=1)
    quantity: int = Field(..., ge=1)
    reservation_id: int | None = Field(default=None, ge=1)


class TransferReceptionItem(BaseModel):
    item_id: int = Field(
        ...,
        ge=1,
        validation_alias=AliasChoices("item_id", "device_id"),
    )
    received_quantity: int = Field(
        default=0,
        ge=0,
        validation_alias=AliasChoices("received_quantity", "quantity"),
    )


class TransferOrderItemCreate(TransferOrderItemBase):
    """Elemento incluido en la creación de una orden de transferencia."""


class TransferOrderTransition(BaseModel):
    reason: str | None = Field(default=None, max_length=255)
    items: list[TransferReceptionItem] | None = None


class TransferOrderCreate(BaseModel):
    origin_store_id: int = Field(..., ge=1)
    destination_store_id: int = Field(..., ge=1)
    reason: str | None = Field(default=None, max_length=255)
    items: list[TransferOrderItemCreate]

    @field_validator("reason")
    @classmethod
    def _normalize_reason(cls, value: str | None) -> str | None:
        if value is None:
            return value
        normalized = value.strip()
        return normalized or None

    @field_validator("items")
    @classmethod
    def _ensure_items(cls, value: list[TransferOrderItemCreate]) -> list[TransferOrderItemCreate]:
        if not value:
            raise ValueError(
                "Debes incluir al menos un dispositivo en la transferencia.")
        return value


class TransferOrderItemResponse(TransferOrderItemBase):
    id: int
    transfer_order_id: int
    dispatched_quantity: int
    received_quantity: int
    dispatched_unit_cost: Decimal | None = Field(default=None, ge=0)
    device_name: str
    sku: str

    model_config = ConfigDict(from_attributes=True)


class TransferOrderResponse(BaseModel):
    id: int
    origin_store_id: int
    destination_store_id: int
    origin_store_name: str
    destination_store_name: str
    status: TransferStatus
    reason: str | None
    created_at: datetime
    updated_at: datetime
    dispatched_at: datetime | None
    received_at: datetime | None
    cancelled_at: datetime | None
    items: list[TransferOrderItemResponse]
    ultima_accion: AuditTrailInfo | None = None

    model_config = ConfigDict(from_attributes=True)


class TransferReportFilters(BaseModel):
    store_id: int | None = None
    origin_store_id: int | None = None
    destination_store_id: int | None = None
    status: TransferStatus | None = None
    date_from: datetime | None = None
    date_to: datetime | None = None


class TransferReportDevice(BaseModel):
    sku: str | None
    name: str | None
    quantity: int


class TransferReportItem(BaseModel):
    id: int
    folio: str
    origin_store: str
    destination_store: str
    status: TransferStatus
    reason: str | None
    requested_at: datetime
    dispatched_at: datetime | None
    received_at: datetime | None
    cancelled_at: datetime | None
    requested_by: str | None
    dispatched_by: str | None
    received_by: str | None
    cancelled_by: str | None
    total_quantity: int
    devices: list[TransferReportDevice]
    ultima_accion: AuditTrailInfo | None = None


class TransferReportTotals(BaseModel):
    total_transfers: int
    pending: int
    in_transit: int
    completed: int
    cancelled: int
    total_quantity: int


class TransferReport(BaseModel):
    generated_at: datetime
    filters: TransferReportFilters
    totals: TransferReportTotals
    items: list[TransferReportItem]

File: frontend/src/api/transfers.ts
import { request, requestCollection } from "./client";

export type TransferOrderItem = {
  id: number;
  transfer_order_id: number;
  device_id: number;
  quantity: number;
  reservation_id?: number | null;
  device_name?: string | null;
  sku?: string | null;
  dispatched_quantity?: number;
  received_quantity?: number;
};

export type TransferOrder = {
  id: number;
  origin_store_id: number;
  destination_store_id: number;
  status: "SOLICITADA" | "EN_TRANSITO" | "RECIBIDA" | "CANCELADA";
  reason?: string | null;
  created_at: string;
  updated_at: string;
  dispatched_at?: string | null;
  received_at?: string | null;
  cancelled_at?: string | null;
  items: TransferOrderItem[];
  origin_store_name?: string | null;
  destination_store_name?: string | null;
  ultima_accion?: {
    usuario?: string | null;
    timestamp: string;
  } | null;
};

export type TransferOrderInput = {
  origin_store_id: number;
  destination_store_id: number;
  reason?: string;
  items: { device_id: number; quantity: number; reservation_id?: number | null }[];
};

export type TransferReceptionItem = {
  item_id: number;
  received_quantity: number;
};

export type TransferTransitionInput = {
  reason?: string;
  items?: TransferReceptionItem[];
};

export type TransferReportDevice = {
  sku?: string | null;
  name?: string | null;
  quantity: number;
};

export type TransferReportItem = {
  id: number;
  folio: string;
  origin_store: string;
  destination_store: string;
  status: TransferOrder["status"];
  reason?: string | null;
  requested_at: string;
  dispatched_at?: string | null;
  received_at?: string | null;
  cancelled_at?: string | null;
  requested_by?: string | null;
  dispatched_by?: string | null;
  received_by?: string | null;
  cancelled_by?: string | null;
  total_quantity: number;
  devices: TransferReportDevice[];
};

export type TransferReportTotals = {
  total_transfers: number;
  pending: number;
  in_transit: number;
  completed: number;
  cancelled: number;
  total_quantity: number;
};

export type TransferReportFilters = {
  store_id?: number;
  origin_store_id?: number;
  destination_store_id?: number;
  status?: TransferOrder["status"];
  date_from?: string;
  date_to?: string;
};

export type TransferReport = {
  generated_at: string;
  filters: TransferReportFilters;
  totals: TransferReportTotals;
  items: TransferReportItem[];
};

export function listTransfers(token: string, storeId?: number): Promise<TransferOrder[]> {
  const params = new URLSearchParams();
  params.set("limit", "25");
  if (typeof storeId === "number") {
    params.set("store_id", String(storeId));
  }
  const query = params.toString();
  const path = `/transfers${query ? `?${query}` : ""}`;
  return requestCollection<TransferOrder>(path, { method: "GET" }, token);
}

export function getTransfer(token: string, transferId: number): Promise<TransferOrder> {
  return request<TransferOrder>(`/transfers/${transferId}`, { method: "GET" }, token);
}

export function createTransferOrder(
  token: string,
  payload: TransferOrderInput,
  reason: string
): Promise<TransferOrder> {
  return request(
    "/transfers",
    { method: "POST", body: JSON.stringify(payload), headers: { "X-Reason": reason } },
    token
  );
}

export function dispatchTransferOrder(
  token: string,
  transferId: number,
  payload: TransferTransitionInput,
  reason: string
): Promise<TransferOrder> {
  return request(
    `/transfers/${transferId}/dispatch`,
    {
      method: "POST",
      body: JSON.stringify(payload),
      headers: { "X-Reason": reason },
    },
    token
  );
}

export function receiveTransferOrder(
  token: string,
  transferId: number,
  payload: TransferTransitionInput,
  reason: string
): Promise<TransferOrder> {
  return request(
    `/transfers/${transferId}/receive`,
    {
      method: "POST",
      body: JSON.stringify(payload),
      headers: { "X-Reason": reason },
    },
    token
  );
}

export function cancelTransferOrder(
  token: string,
  transferId: number,
  payload: TransferTransitionInput,
  reason: string
): Promise<TransferOrder> {
  return request(
    `/transfers/${transferId}/cancel`,
    {
      method: "POST",
      body: JSON.stringify(payload),
      headers: { "X-Reason": reason },
    },
    token
  );
}

function buildTransferReportQuery(filters: TransferReportFilters = {}): string {
  const params = new URLSearchParams();
  if (typeof filters.store_id === "number") {
    params.set("store_id", String(filters.store_id));
  }
  if (typeof filters.origin_store_id === "number") {
    params.set("origin_store_id", String(filters.origin_store_id));
  }
  if (typeof filters.destination_store_id === "number") {
    params.set("destination_store_id", String(filters.destination_store_id));
  }
  if (filters.status) {
    params.set("status", filters.status);
  }
  if (filters.date_from) {
    params.set("date_from", filters.date_from);
  }
  if (filters.date_to) {
    params.set("date_to", filters.date_to);
  }
  const query = params.toString();
  return query ? `?${query}` : "";
}

export function getTransferReport(
  token: string,
  filters: TransferReportFilters = {},
): Promise<TransferReport> {
  const query = buildTransferReportQuery(filters);
  return request<TransferReport>(`/transfers/report${query}`, { method: "GET" }, token);
}

export function exportTransferReportPdf(
  token: string,
  reason: string,
  filters: TransferReportFilters = {},
): Promise<Blob> {
  const query = buildTransferReportQuery(filters);
  return request<Blob>(
    `/transfers/export/pdf${query}`,
    { method: "GET", headers: { "X-Reason": reason }, responseType: "blob" },
    token,
  );
}

export function exportTransferReportExcel(
  token: string,
  reason: string,
  filters: TransferReportFilters = {},
): Promise<Blob> {
  const query = buildTransferReportQuery(filters);
  return request<Blob>(
    `/transfers/export/xlsx${query}`,
    { method: "GET", headers: { "X-Reason": reason }, responseType: "blob" },
    token,
  );
}
